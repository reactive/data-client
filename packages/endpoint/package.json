{
  "name": "@data-client/endpoint",
  "version": "0.14.17",
  "description": "Declarative Network Interface Definitions",
  "homepage": "https://dataclient.io/docs/guides/custom-protocol",
  "keywords": [
    "data",
    "cache",
    "networking",
    "fetch",
    "typescript",
    "data fetching",
    "data cache",
    "api",
    "normalized cache",
    "subject pattern",
    "declarative",
    "definition",
    "schema",
    "async",
    "ajax",
    "react",
    "front-end",
    "mobile",
    "react-native",
    "ios",
    "android",
    "web",
    "expo",
    "async",
    "promise"
  ],
  "sideEffects": false,
  "main": "dist/index.js",
  "react-native": "legacy/index.js",
  "module": "legacy/index.js",
  "unpkg": "dist/index.umd.min.js",
  "types": "lib/index.d.ts",
  "typesVersions": {
    ">=5.4": {
      "": [
        "lib/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=4.8": {
      "": [
        "ts4.8/index.d.ts"
      ],
      "*": [
        "ts4.8/index.d.ts"
      ]
    },
    ">=4.2": {
      "": [
        "ts4.2/index.d.ts"
      ],
      "*": [
        "ts4.2/index.d.ts"
      ]
    },
    ">=4.0": {
      "": [
        "ts4.0/index.d.ts"
      ],
      "*": [
        "ts4.0/index.d.ts"
      ]
    },
    ">=3.4": {
      "": [
        "ts3.4/index.d.ts"
      ],
      "*": [
        "ts3.4/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": {
      "module": "./lib/index.js",
      "import": "./node.mjs",
      "require": "./dist/index.js",
      "default": "./lib/index.js"
    },
    "./package.json": "./package.json"
  },
  "type": "module",
  "engines": {
    "node": "^12.17 || ^13.7 || >=14"
  },
  "files": [
    "src",
    "dist",
    "lib",
    "ts4.8",
    "ts4.2",
    "ts4.0",
    "ts3.4",
    "node.mjs",
    "legacy",
    "LICENSE",
    "README.md"
  ],
  "scripts": {
    "build:lib": "NODE_ENV=production BROWSERSLIST_ENV='2020' yarn g:babel --out-dir lib --ignore 'src/schemas-3.7'",
    "build:legacy:lib": "NODE_ENV=production BROWSERSLIST_ENV='2018' yarn g:babel --out-dir legacy --ignore 'src/schemas-3.7'",
    "build:js:node": "BROWSERSLIST_ENV=node12 yarn g:rollup && echo '{\"type\":\"commonjs\"}' > dist/package.json",
    "build:js:browser": "BROWSERSLIST_ENV=legacy yarn g:rollup",
    "build:bundle": "yarn g:runs build:js:\\*",
    "build:clean": "yarn g:clean ts3.4 ts4.0 ts4.2 ts4.8",
    "build": "run build:lib && run build:legacy:lib && run build:bundle",
    "build:legacy-types": "yarn g:legacy-types 4.8 4.2 4.0 3.4",
    "dev": "run build:lib -w",
    "prepare": "run build:lib",
    "prepack": "run prepare && run build:bundle && run build:legacy:lib",
    "tsc:ci": "yarn g:tsc --project tsconfig.test.json",
    "typecheck": "run tsc:ci"
  },
  "author": "Nathaniel Tucker <me@ntucker.me> (https://github.com/ntucker)",
  "funding": "https://github.com/sponsors/ntucker",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:reactive/data-client.git",
    "directory": "packages/endpoint"
  },
  "bugs": {
    "url": "https://github.com/reactive/data-client/issues"
  },
  "dependencies": {
    "@babel/runtime-corejs3": "^7.26.0",
    "core-js-pure": "^3.40.0"
  },
  "devDependencies": {
    "@anansi/browserslist-config": "^1.4.2",
    "@data-client/core": "workspace:*",
    "@data-client/normalizr": "workspace:*",
    "@js-temporal/polyfill": "^0.4.4",
    "@types/jest": "^29.5.14",
    "@types/node": "^22.0.0",
    "immutable": "4.2.2",
    "nock": "13.3.1",
    "rollup-plugins": "workspace:*"
  }
}
