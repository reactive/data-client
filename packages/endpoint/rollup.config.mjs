import {
  babel,
  dts,
  commonjs,
  filesize,
  json,
  resolve,
  replace,
  terser,
  onwarn,
  resolveTsAsJs,
} from 'rollup-plugins';

import pkg from './package.json' with { type: 'json' };

const dependencies = Object.keys(pkg.dependencies).filter(
  dep => !['@babel/runtime'].includes(dep),
);

const extensions = ['.js', '.ts', '.tsx', '.mjs', '.json', '.node'];
const nativeExtensions = ['.native.ts', ...extensions];
process.env.NODE_ENV = 'production';

function isExternal(id) {
  return dependencies.some(dep => dep === id || id.startsWith(dep));
}

const configs = [];
if (process.env.BROWSERSLIST_ENV !== 'node12') {
  // browser-friendly UMD build
  configs.push({
    input: 'lib/index.js',
    output: [{ file: pkg.unpkg, format: 'umd', name: 'Endpoint' }],
    onwarn,
    plugins: [
      babel({
        exclude: ['node_modules/**', '/**__tests__/**'],
        extensions,
        rootMode: 'upward',
        babelHelpers: 'runtime',
        caller: { polyfillMethod: false },
      }),
      replace({
        'process.env.NODE_ENV': JSON.stringify('production'),
        preventAssignment: true,
      }),
      resolve({ extensions }),
      commonjs({ extensions }),
      json(),
      terser({}),
      filesize({ showBrotliSize: true }),
    ],
  });
  configs.push({
    input: './lib/index.d.ts',
    output: [{ file: 'index.d.ts', format: 'es' }],
    plugins: [dts()],
  });
} else {
  // node-friendly commonjs build
  configs.push({
    input: 'src/index.ts',
    external: isExternal,
    output: [{ file: pkg.main, format: 'cjs' }],
    onwarn,
    plugins: [
      babel({
        exclude: ['node_modules/**', '**/__tests__/**', '**/*.d.ts'],
        extensions: nativeExtensions,
        rootMode: 'upward',
        babelHelpers: 'runtime',
      }),
      replace({ 'process.env.CJS': 'true', preventAssignment: true }),
      resolve({ extensions: nativeExtensions }),
      resolveTsAsJs,
      commonjs({ extensions: nativeExtensions }),
    ],
  });
}
export default configs;
