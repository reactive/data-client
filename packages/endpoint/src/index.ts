export type {
  EndpointOptions,
  EndpointInstance,
  EndpointInstanceInterface,
  EndpointExtendOptions,
} from './endpoint.js';
export * as schema from './schema.js';
// Without this we get 'cannot be named without a reference to' for resource()....why is this?
// Clue 1) It only happens with types mentioned in return types of other types
export type { Array, Invalidate, Collection, DefaultArgs } from './schema.js';
export { default as Entity } from './schemas/Entity.js';
export { default as EntityMixin } from './schemas/EntityMixin.js';
export { default as validateRequired } from './validateRequired.js';
export { INVALID } from './special.js';
export type {
  EndpointInterface,
  ReadEndpoint,
  MutateEndpoint,
  Schema,
  SnapshotInterface,
  ExpiryStatusInterface,
  SchemaSimple,
  SchemaClass,
  PolymorphicInterface,
  Queryable,
} from './interface.js';
export type { EntityFields } from './schemas/EntityFields.js';
export type {
  AbstractInstanceType,
  Normalize,
  NormalizeNullable,
  Denormalize,
  DenormalizeNullable,
  SchemaArgs,
  ObjectArgs,
  EntityMap,
  RecordClass,
  NormalizedNullableObject,
  NormalizeObject,
  NormalizedEntity,
  DenormalizeObject,
  DenormalizeNullableObject,
} from './normal.js';
export type {
  EndpointExtraOptions,
  FetchFunction,
  ResolveType,
  EndpointParam,
  NetworkError,
  UnknownError,
  ErrorTypes,
  EndpointToFunction,
} from './types.js';
export type { NI } from './NoInfer.js';

export { default as Endpoint, ExtendableEndpoint } from './endpoint.js';
export type { KeyofEndpointInstance } from './endpoint.js';
