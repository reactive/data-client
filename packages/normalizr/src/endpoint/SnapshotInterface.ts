import type { EndpointInterface } from './EndpointInterface.js';
import type { ErrorTypes } from './ErrorTypes.js';
import type { ExpiryStatusInterface } from '../Expiry.js';
import { Queryable } from '../interface.js';
import { DenormalizeNullable } from '../types.js';

export interface SnapshotInterface {
  /**
   * Gets the (globally referentially stable) response for a given endpoint/args pair from state given.
   * @see https://dataclient.io/docs/api/Snapshot#getResponse
   */
  getResponse<
    E extends Pick<EndpointInterface, 'key' | 'schema' | 'invalidIfStale'>,
  >(
    endpoint: E,
    ...args: readonly any[]
  ): {
    data: DenormalizeNullable<E['schema']>;
    expiryStatus: ExpiryStatusInterface;
    expiresAt: number;
  };

  /**
   * Gets the (globally referentially stable) response for a given endpoint/args pair from state given.
   * @see https://dataclient.io/docs/api/Snapshot#getResponseMeta
   */
  getResponseMeta<
    E extends Pick<EndpointInterface, 'key' | 'schema' | 'invalidIfStale'>,
  >(
    endpoint: E,
    ...args: readonly any[]
  ): {
    data: DenormalizeNullable<E['schema']>;
    expiryStatus: ExpiryStatusInterface;
    expiresAt: number;
  };

  /** @see https://dataclient.io/docs/api/Snapshot#getError */
  getError: <
    E extends Pick<EndpointInterface, 'key'>,
    Args extends readonly [...Parameters<E['key']>],
  >(
    endpoint: E,
    ...args: Args
  ) => ErrorTypes | undefined;

  /**
   * Retrieved memoized value for any Querable schema
   * @see https://dataclient.io/docs/api/Snapshot#get
   */
  get<S extends Queryable>(schema: S, ...args: readonly any[]): any;

  /**
   * Queries the store for a Querable schema; providing related metadata
   * @see https://dataclient.io/docs/api/Snapshot#getQueryMeta
   */
  getQueryMeta<S extends Queryable>(
    schema: S,
    ...args: readonly any[]
  ): {
    data: any;
    countRef: () => () => void;
  };

  readonly fetchedAt: number;
  readonly abort: Error;
}
