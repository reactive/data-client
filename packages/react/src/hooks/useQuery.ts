import type {
  DenormalizeNullable,
  NI,
  Queryable,
  SchemaArgs,
} from '@data-client/core';
import { useEffect, useMemo } from 'react';

import useCacheState from './useCacheState.js';
import useController from './useController.js';

/**
 * Query the store.
 *
 * `useQuery` results are globally memoized.
 * @see https://dataclient.io/docs/api/useQuery
 */
export default function useQuery<S extends Queryable>(
  schema: S,
  ...args: NI<SchemaArgs<S>>
  // INVALID results in undefined regardless of DenormalizeNullable
): DenormalizeNullable<S> | undefined {
  const state = useCacheState();
  const controller = useController();

  const key = JSON.stringify(args);

  // even though controller.get() is memoized, its memoization is more complex than
  // this so we layer it up to improve rerenders
  const { data, countRef } = useMemo(() => {
    return controller.getQueryMeta(schema, ...args, state);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [state.indexes, state.entities, key]);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(countRef, [data]);

  return data;
}
