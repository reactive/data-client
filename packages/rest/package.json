{
  "name": "@data-client/rest",
  "version": "0.14.18",
  "description": "Quickly define typed REST resources and endpoints",
  "homepage": "https://dataclient.io/rest",
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:reactive/data-client.git",
    "directory": "packages/rest"
  },
  "bugs": {
    "url": "https://github.com/reactive/data-client/issues"
  },
  "keywords": [
    "REST",
    "CRUD",
    "networking",
    "fetch",
    "data",
    "cache",
    "model",
    "resource",
    "typescript",
    "HTTP",
    "path-to-regexp",
    "path template",
    "data fetching",
    "data cache",
    "api",
    "normalized cache",
    "subject pattern",
    "endpoint",
    "mutation",
    "front-end",
    "mobile",
    "react-native",
    "ios",
    "android",
    "web",
    "json",
    "async",
    "promise"
  ],
  "sideEffects": false,
  "main": "dist/index.js",
  "react-native": "legacy/index.js",
  "module": "legacy/index.js",
  "unpkg": "dist/index.umd.min.js",
  "types": "lib/index.d.ts",
  "typesVersions": {
    ">=5.0": {
      "": [
        "lib/index.d.ts"
      ],
      "next": [
        "lib/next/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=4.1": {
      "": [
        "ts4.1/index.d.ts"
      ],
      "next": [
        "ts4.1/next/index.d.ts"
      ],
      "*": [
        "ts4.1/index.d.ts"
      ]
    },
    ">=4.0": {
      "": [
        "ts4.0/index.d.ts"
      ],
      "next": [
        "ts4.0/next/index.d.ts"
      ],
      "*": [
        "ts4.0/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": {
      "module": "./lib/index.js",
      "import": "./node.mjs",
      "require": "./dist/index.js",
      "default": "./lib/index.js"
    },
    "./next": {
      "types": "./lib/next/index.d.ts",
      "require": "./dist/next.js",
      "default": "./lib/next/index.js"
    },
    "./package.json": "./package.json"
  },
  "type": "module",
  "engines": {
    "node": "^12.17 || ^13.7 || >=14"
  },
  "files": [
    "src",
    "dist",
    "lib",
    "node.mjs",
    "legacy",
    "ts4.0",
    "ts4.1",
    "LICENSE",
    "README.md"
  ],
  "scripts": {
    "build:lib": "NODE_ENV=production BROWSERSLIST_ENV='2020' yarn g:babel --out-dir lib",
    "build:legacy:lib": "NODE_ENV=production BROWSERSLIST_ENV='2018' yarn g:babel --out-dir legacy",
    "build:js:node": "BROWSERSLIST_ENV=node12 yarn g:rollup",
    "build:js:browser": "BROWSERSLIST_ENV=legacy yarn g:rollup",
    "build:bundle": "yarn g:runs build:js:\\* && echo '{\"type\":\"commonjs\"}' > dist/package.json",
    "build:clean": "yarn g:clean ts4.0 ts4.1 index.d.ts next.d.ts",
    "build:legacy-types": "yarn g:downtypes lib ts4.0 --to=4.0 && yarn g:downtypes lib ts4.1 --to=4.1 && yarn g:copy --up 1 ./src-4.1-types/**/*.d.ts ./ts4.0/ && yarn g:copy --up 1 ./src-4.1-types/**/*.d.ts ./ts4.1 && yarn g:copy --up 1 ./src-4.0-types/**/*.d.ts ./ts4.0",
    "build": "run build:lib && run build:legacy:lib && run build:bundle",
    "dev": "run build:lib -w",
    "prepare": "run build:lib",
    "prepack": "run prepare && run build:bundle && run build:legacy:lib",
    "tsc:ci": "yarn g:tsc --project tsconfig.test.json",
    "typecheck": "run tsc:ci"
  },
  "author": "Nathaniel Tucker <me@ntucker.me> (https://github.com/ntucker)",
  "funding": "https://github.com/sponsors/ntucker",
  "license": "Apache-2.0",
  "dependencies": {
    "@babel/runtime-corejs3": "^7.26.0",
    "@data-client/endpoint": "^0.14.17",
    "core-js-pure": "^3.40.0",
    "path-to-regexp": "^6.3.0"
  },
  "devDependencies": {
    "@anansi/browserslist-config": "^1.4.2",
    "@data-client/react": "workspace:*",
    "@data-client/test": "workspace:*",
    "@types/jest": "^29.5.14",
    "@types/node": "^22.0.0",
    "nock": "13.3.1",
    "rollup-plugins": "workspace:*"
  }
}
