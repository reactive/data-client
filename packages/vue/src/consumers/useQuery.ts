import type {
  DenormalizeNullable,
  NI,
  Queryable,
  SchemaArgs,
} from '@data-client/core';
import { computed, watch, type ComputedRef } from 'vue';

import { useController, injectState } from '../context.js';

/**
 * Query the store (non-suspense).
 *
 * Returns a readonly computed ref of the query result. The value is undefined when
 * the result is not found or invalid.
 * Mirrors React's useQuery semantics using Vue reactivity.
 * @see https://dataclient.io/docs/api/useQuery
 */
export default function useQuery<S extends Queryable>(
  schema: S,
  ...args: NI<SchemaArgs<S>>
): ComputedRef<DenormalizeNullable<S> | undefined> {
  const stateRef = injectState();
  const controller = useController();

  // Compute query meta based on state and args. This mirrors React's memoization
  // that keys off state.entities/indexes and args.
  const queryMeta = computed(() =>
    controller.getQueryMeta(schema, ...args, stateRef.value as any),
  );

  // Maintain GC refcounts on data mount/changes
  watch(
    () => queryMeta.value.data,
    (_newVal, _oldVal, onCleanup) => {
      const decrement = queryMeta.value.countRef();
      onCleanup(() => decrement());
    },
    { immediate: true },
  );

  return computed(() => queryMeta.value.data);
}
