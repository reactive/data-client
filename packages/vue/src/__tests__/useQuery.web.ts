import { schema } from '@data-client/endpoint';
import { resource } from '@data-client/rest';

import {
  ArticleWithSlug,
  ArticleSlugResource,
  ArticleResource,
  Article,
  IDEntity,
  UnionSchema,
  FirstUnion,
  SecondUnion,
} from '../../../../__tests__/new';
import useQuery from '../consumers/useQuery';
import { renderDataComposable } from '../test';

// Inline fixtures (duplicated from React tests to avoid cross-project imports)
const payloadSlug = {
  id: 5,
  title: 'hi ho',
  slug: 'hi-ho',
  content: 'whatever',
  tags: ['a', 'best', 'react'],
};

const nested = [
  {
    id: 5,
    title: 'hi ho',
    content: 'whatever',
    tags: ['a', 'best', 'react'],
    author: {
      id: 23,
      username: 'bob',
    },
  },
  {
    id: 3,
    title: 'the next time',
    content: 'whatever',
    author: {
      id: 23,
      username: 'charles',
      email: 'bob@bob.com',
    },
  },
];

describe('vue useQuery()', () => {
  it('should be undefined with empty state', () => {
    const { result } = renderDataComposable(() => {
      return useQuery(ArticleWithSlug, { id: payloadSlug.id });
    }, {});
    // @ts-expect-error
    result.current?.value?.doesnotexist;
    expect(result.current?.value).toBe(undefined);
  });

  it('All should be undefined with empty state', () => {
    const { result } = renderDataComposable(() => {
      return useQuery(new schema.All(ArticleWithSlug));
    }, {});
    // @ts-expect-error
    result.current?.value?.doesnotexist;
    expect(result.current?.value).toBe(undefined);
  });

  it('should fail on schema.Array', () => {
    const { result } = renderDataComposable(() => {
      // @ts-expect-error
      return useQuery(new schema.Array(ArticleWithSlug));
    }, {});
    // @ts-expect-error
    result.current?.value?.doesnotexist;
    expect(result.current?.value).toBe(undefined);
  });

  it('should find Entity by pk', async () => {
    const { result } = renderDataComposable(
      () => {
        return useQuery(ArticleWithSlug, { id: payloadSlug.id });
      },
      {
        initialFixtures: [
          {
            endpoint: ArticleSlugResource.get,
            args: [{ id: 5 }],
            response: payloadSlug,
          },
        ],
      },
    );
    expect(result.current?.value).toEqual(ArticleWithSlug.fromJS(payloadSlug));

    // @ts-expect-error
    () => useQuery(ArticleWithSlug);
    // @ts-expect-error
    () => useQuery(ArticleWithSlug, 5, 10);
    // @ts-expect-error
    () => useQuery(ArticleWithSlug, 5);
    // @ts-expect-error
    () => useQuery(ArticleWithSlug, { id5: 5 });
    // @ts-expect-error
    () => useQuery(ArticleWithSlug, '5');
  });

  it('should find Entity by slug', async () => {
    const { result } = renderDataComposable(
      () => {
        return useQuery(ArticleWithSlug, { slug: payloadSlug.slug });
      },
      {
        initialFixtures: [
          {
            endpoint: ArticleSlugResource.get,
            args: [{ id: 5 }],
            response: payloadSlug,
          },
        ],
      },
    );
    expect(result.current?.value).toEqual(ArticleWithSlug.fromJS(payloadSlug));
  });

  it('should select Collections', () => {
    const initialFixtures = [
      {
        endpoint: ArticleResource.getList,
        args: [],
        response: nested,
      },
    ];
    const { result } = renderDataComposable(
      () => {
        return useQuery(ArticleResource.getList.schema);
      },
      { initialFixtures },
    );
    expect(result.current?.value).toBeDefined();
    if (!result.current?.value) return;
    expect(result.current.value.length).toBe(nested.length);
    expect(result.current.value[0]).toBeInstanceOf(Article);
    expect(result.current.value).toMatchSnapshot();
  });

  it('should update Collections when pushed', async () => {
    const initialFixtures = [
      {
        endpoint: ArticleResource.getList,
        args: [],
        response: nested,
      },
    ];
    const resolverFixtures = [
      {
        endpoint: ArticleResource.getList.push,
        response(body: any) {
          return body;
        },
      },
    ];
    const { result, controller, waitForNextUpdate } = renderDataComposable(
      () => {
        return useQuery(ArticleResource.getList.schema, {});
      },
      { initialFixtures, resolverFixtures },
    );
    expect(result.current?.value).toBeDefined();
    if (!result.current?.value) return;
    expect(result.current.value.length).toBe(nested.length);
    await controller.fetch(ArticleResource.getList.push, {
      id: 50,
      title: 'newly added',
      content: 'this one is pushed',
    });
    await waitForNextUpdate();
    expect(result.current.value.length).toBe(nested.length + 1);
  });

  it('should retrieve a nested collection', () => {
    class Todo extends IDEntity {
      userId = 0;
      title = '';
      completed = false;

      static key = 'Todo';
    }

    class User extends IDEntity {
      name = '';
      username = '';
      email = '';
      todos: Todo[] = [];

      static key = 'User';
      static schema = {
        todos: new schema.Collection(new schema.Array(Todo), {
          nestKey: parent => ({
            userId: parent.id,
          }),
        }),
      };
    }

    const userTodos = new schema.Collection(new schema.Array(Todo), {
      argsKey: ({ userId }: { userId: string }) => ({
        userId,
      }),
    });

    const UserResource = resource({ schema: User, path: '/users/:id' });

    const initialFixtures = [
      {
        endpoint: UserResource.get,
        args: [{ id: '1' }],
        response: {
          id: '1',
          todos: [
            {
              id: '5',
              title: 'finish collections',
              userId: '1',
            },
          ],
          username: 'bob',
        },
      },
    ];
    const { result } = renderDataComposable(
      () => {
        return useQuery(userTodos, { userId: '1' });
      },
      { initialFixtures },
    );
    expect(result.current?.value).toBeDefined();
    if (!result.current?.value) return;
    expect(result.current.value.length).toBe(1);
    expect(result.current.value[0]).toBeInstanceOf(Todo);
    expect(result.current.value).toMatchSnapshot();

    // @ts-expect-error
    () => useQuery(userTodos, { userId: { hi: '5' } });
    // @ts-expect-error
    () => useQuery(userTodos, { userIdd: '1' });
    // @ts-expect-error
    () => useQuery(userTodos, { bob: '1' });
    // @ts-expect-error
    () => useQuery(userTodos);
  });

  it('should work with unions', async () => {
    const warnSpy = jest
      .spyOn(global.console, 'warn')
      .mockImplementation(() => {
        // noop
      });
    const UnionResource = resource({
      path: '/union/:id',
      schema: UnionSchema,
    });

    const { result } = renderDataComposable(
      () => {
        return useQuery(UnionResource.get.schema, { id: '5', type: 'first' });
      },
      {
        initialFixtures: [
          {
            endpoint: UnionResource.getList,
            args: [],
            response: [
              { id: '5', body: 'hi', type: 'first' },
              { id: '6', body: 'hi', type: 'another' },
            ],
          },
        ],
      },
    );
    expect(result.current?.value).toBeDefined();
    if (!result.current?.value) {
      warnSpy.mockRestore();
      return;
    }
    expect(result.current.value.type).toBe('first');
    expect(result.current.value).toBeInstanceOf(FirstUnion);

    // these are the 'fallback case' where it cannot determine type discriminator, so just enumerates
    () => useQuery(UnionResource.get.schema, { id: '5' });
    () => useQuery(UnionResource.get.schema, { body: '5' });

    // @ts-expect-error
    () => useQuery(UnionResource.get.schema, { id: '5', type: 'notvalid' });
    // @ts-expect-error
    () => useQuery(UnionResource.get.schema, { doesnotexist: '5' });
    // @ts-expect-error
    () => useQuery(UnionResource.get.schema);

    expect(warnSpy.mock.calls).toMatchSnapshot();
    warnSpy.mockRestore();
  });

  it('should work with unions collections', async () => {
    const warnSpy = jest
      .spyOn(global.console, 'warn')
      .mockImplementation(() => {
        // noop
      });

    const UnionResource = resource({
      path: '/union/:id',
      schema: UnionSchema,
    });

    const { result, controller, waitForNextUpdate } = renderDataComposable(
      () => {
        return useQuery(UnionResource.getList.schema);
      },
      {
        initialFixtures: [
          {
            endpoint: UnionResource.getList,
            args: [],
            response: [
              undefined,
              null,
              { id: '5', body: 'hi', type: 'first' },
              { id: '6', body: 'hi', type: 'another' },
              { id: '7', body: 'hi' },
            ],
          },
        ],
        resolverFixtures: [
          {
            endpoint: UnionResource.getList.push,
            response(body) {
              return body;
            },
          },
        ],
      },
    );
    expect(result.current?.value).toBeDefined();
    if (!result.current?.value) return;
    expect(result.current.value[0]).toBeNull();
    expect(result.current.value[1]).toBeInstanceOf(FirstUnion);
    expect(result.current.value[2]).not.toBeInstanceOf(FirstUnion);
    expect(result.current.value[3]).not.toBeInstanceOf(FirstUnion);
    expect(warnSpy.mock.calls).toMatchSnapshot();

    await controller.fetch(UnionResource.getList.push, {
      body: 'hi',
      type: 'second',
      id: '100',
    });
    await waitForNextUpdate();
    expect(result.current.value[4]).toBeInstanceOf(SecondUnion);
    expect(result.current.value).toMatchSnapshot();
    warnSpy.mockRestore();
  });
});
