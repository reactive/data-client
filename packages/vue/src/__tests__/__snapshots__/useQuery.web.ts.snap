// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`vue useQuery() should retrieve a nested collection 1`] = `
[
  Todo {
    "completed": false,
    "id": "5",
    "title": "finish collections",
    "userId": "1",
  },
]
`;

exports[`vue useQuery() should select Collections 1`] = `
[
  Article {
    "author": User {
      "email": "bob@bob.com",
      "id": 23,
      "isAdmin": false,
      "username": "charles",
    },
    "content": "whatever",
    "id": 5,
    "tags": [
      "a",
      "best",
      "react",
    ],
    "title": "hi ho",
  },
  Article {
    "author": User {
      "email": "bob@bob.com",
      "id": 23,
      "isAdmin": false,
      "username": "charles",
    },
    "content": "whatever",
    "id": 3,
    "tags": [],
    "title": "the next time",
  },
]
`;

exports[`vue useQuery() should work with unions 1`] = `
[
  [
    "Schema attribute "another" is not expected.
Expected one of: "first", "second"

Value: {
  "id": "6",
  "body": "hi",
  "type": "another"
}",
  ],
]
`;

exports[`vue useQuery() should work with unions collections 1`] = `
[
  [
    "Schema attribute "another" is not expected.
Expected one of: "first", "second"

Value: {
  "id": "6",
  "body": "hi",
  "type": "another"
}",
  ],
  [
    "Schema attribute undefined is not expected.
Expected one of: "first", "second"

Value: {
  "id": "7",
  "body": "hi"
}",
  ],
  [
    "TypeError: Unable to determine schema for ArraySchema
Value: {
  "id": "6",
  "body": "hi",
  "type": "another"
}.",
  ],
  [
    "TypeError: Unable to determine schema for ArraySchema
Value: {
  "id": "7",
  "body": "hi"
}.",
  ],
]
`;

exports[`vue useQuery() should work with unions collections 2`] = `
[
  null,
  FirstUnion {
    "body": "hi",
    "firstOnlyField": 5,
    "id": "5",
    "type": "first",
  },
  {
    "body": "hi",
    "id": "6",
    "type": "another",
  },
  {
    "body": "hi",
    "id": "7",
  },
  SecondUnion {
    "body": "hi",
    "id": "100",
    "secondeOnlyField": 10,
    "type": "second",
  },
]
`;
